instance/models/openedx_appserver.py

```diff
@@ -279,6 +298,45 @@ def github_admin_username_list(self):
         admin_users += self.github_admin_users
         return admin_users
 
+    @property
+    def security_groups(self):
+        """
+        List of the names of each OpenStack network security group used by this
+        AppServer
+
+        Example return value: ["edxapp-appserver"]
+        """
+        return [settings.EDXAPP_APPSERVER_SECURITY_GROUP_NAME] + self.additional_security_groups
+
+    def check_security_groups(self):
+        """
+        For security reasons, every edxapp AppServer should be in a security
+        group that only allows access to a few ports, like 443 and 22.
+
+        The security group with the name specified by
+        settings.EDXAPP_APPSERVER_SECURITY_GROUP_NAME is created and managed
+        by this code.
+        """
+        self.logger.info('Checking security groups (OpenStack firewall settings)')
+        network = get_openstack_connection().network
+        main_security_group = network.find_security_group(settings.EDXAPP_APPSERVER_SECURITY_GROUP_NAME)
+        if not main_security_group:
+            # We need to create this security group:
+            main_security_group = network.create_security_group(name=settings.EDXAPP_APPSERVER_SECURITY_GROUP_NAME)
+        description = 'Security group for Open EdX AppServers. Managed automatically by OpenCraft IM.'
+        if main_security_group.description != description:
+            network.update_security_group(main_security_group, description=description)
+
+        # We manage this security group - update its rules to match the configured list of rules
+        sync_security_group_rules(main_security_group, EDXAPP_APPSERVER_SECURITY_GROUP_RULES, network=network)
+
+        # For any additional security groups, just verify that the group exists:
+        groups = self.security_groups
+        groups.remove(main_security_group.name) # We already checked this group
```

Comment:
So that subclasses/mixins can add other security groups by adding their custom groups by overriding `self.security_groups()`, without needing to change this method.